### PR TITLE
add shuffle action

### PR DESCRIPTION
**Pull request by Thor-wl**
ref: https://github.com/volcano-sh/volcano/issues/1955

### CODE REVIEW

File: pkg/scheduler/actions/shuffle/fake_plugin.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package shuffle
+
+import (
+	"volcano.sh/volcano/pkg/scheduler/api"
+	"volcano.sh/volcano/pkg/scheduler/framework"
```
Comment (k82cn) [inline pkg/scheduler/actions/shuffle/fake_plugin.go:21]: let's try [gomock](https://github.com/golang/mock) :)

File: pkg/scheduler/actions/shuffle/shuffle.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+	// select pods that may be evicted
+	tasks := make([]*api.TaskInfo, 0)
+	for _, jobInfo := range ssn.Jobs {
+		for _, taskInfo := range jobInfo.Tasks {
+			if taskInfo.Status == api.Running {
+				tasks = append(tasks, taskInfo)
+			}
+		}
+	}
+	for _, task := range tasks {
```
Comment (k82cn) [inline pkg/scheduler/actions/shuffle/shuffle.go:60]: let's check log level before for looping.

Comment (Thor-wl) [reply pkg/scheduler/actions/shuffle/shuffle.go:60]: OK. Let me check it.

File: pkg/scheduler/actions/shuffle/fake_plugin.go (hunk 1 above)
Comment (Thor-wl) [reply pkg/scheduler/actions/shuffle/fake_plugin.go:21]: Thanks for the advice. Let me take a look at `gomock`.

File: pkg/scheduler/actions/shuffle/shuffle.go (hunk 2 above)
Comment (Thor-wl) [reply pkg/scheduler/actions/shuffle/shuffle.go:60]: Checked and updated.

Comment (Thor-wl) [reply pkg/scheduler/actions/shuffle/shuffle.go:60]: Checked and updated.

File: pkg/scheduler/actions/shuffle/fake_plugin.go (hunk 1 above)
Comment (Thor-wl) [reply pkg/scheduler/actions/shuffle/fake_plugin.go]: Done.

File: pkg/scheduler/framework/mock_gen/action_and_plugin.go [hunk 3]
```diff
@@ -0,0 +1,146 @@
+// Code generated by MockGen. DO NOT EDIT.
```
[the remark this answers is not in the document]
Comment (Thor-wl) [reply pkg/scheduler/framework/mock_gen/action_and_plugin.go:1]: It is generated from `interface.go`.

File: pkg/scheduler/framework/session_plugins.go [hunk 4]
```diff
@@ -444,47 +443,28 @@ func (ssn *Session) TargetJob(jobs []*api.JobInfo) *api.JobInfo {
...
-				for _, v := range victims {
-					for _, c := range candidates {
-						if v.UID == c.UID {
-							intersection = append(intersection, v)
-						}
-					}
+			for _, fn := range fns {
+				victimTasks := fn(tasks)
+				for _, victim := range victimTasks {
+					victimSet[victim] = true
```
[the remark this answers is not in the document]
Comment (Thor-wl) [reply pkg/scheduler/framework/session_plugins.go:462]: That's reasonable. Let me update for that.

File: pkg/scheduler/plugins/tdm/tdm.go [hunk 5]
```diff
@@ -291,10 +291,12 @@ func (tp *tdmPlugin) OnSessionOpen(ssn *framework.Session) {
 		return len(jobInfo.TaskStatusIndex[api.Pending]) > 0
 	}
 
+	victimsFns := make([]api.VictimTasksFn, 0)
+	victimsFns = append(victimsFns, victimsFn)
 	ssn.AddPredicateFn(tp.Name(), predicateFn)
 	ssn.AddNodeOrderFn(tp.Name(), nodeOrderFn)
 	ssn.AddPreemptableFn(tp.Name(), preemptableFn)
-	ssn.AddVictimTasksFns(tp.Name(), victimsFn)
+	ssn.AddVictimTasksFns(tp.Name(), victimsFns)
```
[the remark this answers is not in the document]
Comment (Thor-wl) [reply pkg/scheduler/plugins/tdm/tdm.go:299]: It is necessary for `rescheduling` plugin. In `rescheduling`, users can set more than 1 policy to filter victims. It works just like `predicate` in default scheduler.

File: pkg/scheduler/actions/shuffle/shuffle.go (hunk 2 above)
Comment (k82cn) [reply pkg/scheduler/actions/shuffle/shuffle.go:60]: if log level is 3, we still have to go through all tasks, right?

Comment (Thor-wl) [reply pkg/scheduler/actions/shuffle/shuffle.go:60]: I can remove the log for it is only for debug.

**Review by william-wang — APPROVED** [review]
/lgtm

**Pull request merged**